RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,854 @@
+    }
+  }
+}
+```
+
+that is translated to:
+
+```
+{upload.port.address} = 192.168.1.232
+{upload.port.label} = SSH on my-board (192.168.1.232)
+{upload.port.protocol} = ssh
+{upload.port.protocolLabel} = SSH Network port
+{upload.port.properties.macprefix} = AA:BB:CC
+{upload.port.properties.macaddress} = AA:BB:CC:DD:EE:FF
+{serial.port} = 192.168.1.232                  # for backward compatibility
+```
+
+This configuration, together with protocol selection, allows to remove the hardcoded `network_pattern`, now we can replace the legacy recipe (split into multiple lines for clarity):
+
+```
+tools.bossac.upload.network_pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
+                                    -address {serial.port} -port 65280
+                                    -sketch "{build.path}/{build.project_name}.bin"
+```
+
+with:
+
+```
+tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
+                                 -address {upload.port.address} -port 65280
+                                 -sketch "{build.path}/{build.project_name}.bin"
+```
+
+#### User provided fields
+
+The recipe may require some user-entered fields (like username or password). In this case the recipe must use the special placeholder `{upload.field.FIELD_NAME}` where `FIELD_NAME` must be declared separately in the recipe using the following format:
+
+```
+tools.UPLOAD_RECIPE_ID.upload.field.FIELD_NAME=FIELD_LABEL
+tools.UPLOAD_RECIPE_ID.upload.field.FIELD_NAME.secret=true
+```
+
+`FIELD_LABEL` is the label shown in the dialog to ask the user to enter the value of the field. The `secret` property is optional and it should be set to `true` if the field is a secret (like passwords or token).
+
+Let’s see how a complete example will look like:
+
+```
+tools.arduino_ota.upload.field.username=Username
+tools.arduino_ota.upload.field.password=Password
+tools.arduino_ota.upload.field.password.secret=true
+tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
+                                 -address {upload.port.address} -port 65280
+                                 -username "{upload.field.username}
+                                 -password "{upload.field.password}"
+                                 -sketch "{build.path}/{build.project_name}.bin"
+```
+
+#### Support for upload "without a port" (`no-port` dummy protocol)
+
+Some upload tools already have the port detection builtin so there is no need to specify an upload port (for example the `openocd` tool is often able to autodetect the devices to upload by itself).
+
+To support this particular use case a dummy `no-port` protocol has been reserved:
+
+```
+myboard.upload.tool.no-port=openocd_without_port
+```
+
+The `no-port` upload recipe will be selected when:
+
+- The upload port is not specified
+
+or
+
+- The upload port is specified but the protocol of the selected port doesn't match any of the available upload protocols for a board
```

Comment:
Now that I renamed it `default` it's much more straightforward I think but, I agree, we will found out after the first implementation...

Reply:
I agree that default is indeed more matching with the specified behavior, but I do wonder if having a "default" upload tool actually makes sense. It suggests that it can handle ports with different protocols, which I would expect to never really be the case (since different types of ports, i.e. different protocols specify their address and other info in different formats, so the tool recipes must be matched to that). In that sense, a "no port" *is* more descriptive and explicit, since it names a tool that needs no port data at all (and that can also be enforced by not exposing any such data to the recipe, which no longer makes sense with `default`).

I do think that, ideally, `no-port` entries are really a temporary thing, since with the proper discovery tools, I think all upload tools should be able to select a port explicitly (i.e. you can specify the USB device path to openocd), though I can imagine that in practice that some tools will remain `no-port` indefinitely.

Reply:
The fact is that `no-port`, as specified here, actually **will** forward port information if a port is selected (and that would be the case 99% of the time if you use an IDE, you cannot "unselect" a port once you have selected once even if the port disappears).

Reply:
> The fact is that no-port, as specified here, actually will forward port information 

That's an implementation choice, and I would think it would be worth considering *not* to forward any port data to a `no-port` entry (to make it clearer what the intent is and to prevent creating situations that do not make complete sense but need to be supported indefinitely later).

> if you use an IDE, you cannot "unselect" a port once you have selected once

This is also an implementation choice. A port menu could have a "No port" entry. Also, the offered port list could be filtered based on the (protocols supported by the) selected port even.

I'm not entirely sure this is the best way forward, but do think that choices made in the current implementation should not limit this RFC.